crypto/crypto.go

```diff
@@ -6,112 +6,120 @@ import (
 	"github.com/relab/hotstuff/consensus"
 )
 
-type base struct {
-	consensus.CryptoImpl
+type crypto struct {
+	mods *consensus.Modules
+	consensus.CryptoBase
 }
 
-// New returns a new base implementation of the Crypto interface. It will use the given CryptoImpl to create and verify
+// New returns a new implementation of the Crypto interface. It will use the given CryptoBase to create and verify
 // signatures.
-func New(impl consensus.CryptoImpl) consensus.Crypto {
-	return base{CryptoImpl: impl}
+func New(impl consensus.CryptoBase) consensus.Crypto {
+	return &crypto{CryptoBase: impl}
 }
 
 // InitConsensusModule gives the module a reference to the Modules object.
 // It also allows the module to set module options using the OptionsBuilder.
-func (base base) InitConsensusModule(mods *consensus.Modules, cfg *consensus.OptionsBuilder) {
-	if mod, ok := base.CryptoImpl.(consensus.Module); ok {
+func (c *crypto) InitConsensusModule(mods *consensus.Modules, cfg *consensus.OptionsBuilder) {
+	c.mods = mods
+	if mod, ok := c.CryptoBase.(consensus.Module); ok {
 		mod.InitConsensusModule(mods, cfg)
 	}
 }
 
 // CreatePartialCert signs a single block and returns the partial certificate.
-func (base base) CreatePartialCert(block *consensus.Block) (cert consensus.PartialCert, err error) {
-	sig, err := base.Sign(block.Hash())
+func (c crypto) CreatePartialCert(block *consensus.Block) (cert consensus.PartialCert, err error) {
+	sig, err := c.Sign(block.ToBytes())
 	if err != nil {
 		return consensus.PartialCert{}, err
```

Comment:
We should avoid returning empty structs when we have error; it causes unnecessary allocation.